mongodbatlas/fw_resource_mongodbatlas_encryption_at_rest.go

```diff
@@ -0,0 +1,576 @@
+		atlasEncryptionAtRest.AzureKeyVault = *newAtlasAzureKeyVault(encryptionAtRestPlan.AzureKeyVaultConfig)
+	}
+	if hasGcpKmsConfigChanged(encryptionAtRestPlan.GoogleCloudKmsConfig, encryptionAtRestState.GoogleCloudKmsConfig) {
+		atlasEncryptionAtRest.GoogleCloudKms = *newAtlasGcpKms(encryptionAtRestPlan.GoogleCloudKmsConfig)
+	}
+
+	atlasEncryptionAtRest.GroupID = projectID
+	_, _, err = conn.EncryptionsAtRest.Create(ctx, atlasEncryptionAtRest)
+	if err != nil {
+		resp.Diagnostics.AddError("error updating encryption at rest", fmt.Sprintf(errorUpdateEncryptionAtRest, err.Error()))
+		return
+	}
+
+	// read
+	encryptionResp, response, err := conn.EncryptionsAtRest.Get(context.Background(), projectID)
+	tflog.Debug(ctx, fmt.Sprintf("encryptionResp from api: %v", encryptionResp))
+	if err != nil {
+		if resp != nil && response.StatusCode == http.StatusNotFound {
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError("error when getting encryption at rest resource after update", fmt.Sprintf(errorReadEncryptionAtRest, err.Error()))
+		return
+	}
+
+	encryptionAtRestStateNew := newTFEncryptionAtRestRSModel(ctx, projectID, encryptionResp, encryptionAtRestPlan)
+	resetDefaultsFromConfig(ctx, encryptionAtRestState, encryptionAtRestStateNew, encryptionAtRestConfig)
+
+	// save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &encryptionAtRestStateNew)...)
+}
+
+func (r *EncryptionAtRestRS) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var encryptionAtRestState *tfEncryptionAtRestRSModel
+
+	// read prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &encryptionAtRestState)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	conn := r.client.Atlas
+	projectID := encryptionAtRestState.ProjectID.ValueString()
+	_, err := conn.EncryptionsAtRest.Delete(ctx, projectID)
+
+	if err != nil {
+		resp.Diagnostics.AddError("error when destroying resource", fmt.Sprintf(errorDeleteEncryptionAtRest, projectID, err.Error()))
+		return
+	}
+}
+
+func (r *EncryptionAtRestRS) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func hasGcpKmsConfigChanged(gcpKmsConfigsPlan, gcpKmsConfigsState []tfGcpKmsConfigModel) bool {
+	return !reflect.DeepEqual(gcpKmsConfigsPlan, gcpKmsConfigsState)
+}
+
+func hasAzureKeyVaultConfigChanged(azureKeyVaultConfigPlan, azureKeyVaultConfigState []tfAzureKeyVaultConfigModel) bool {
+	return !reflect.DeepEqual(azureKeyVaultConfigPlan, azureKeyVaultConfigState)
+}
+
+func hasAwsKmsConfigChanged(awsKmsConfigPlan, awsKmsConfigState []tfAwsKmsConfigModel) bool {
+	return !reflect.DeepEqual(awsKmsConfigPlan, awsKmsConfigState)
+}
+
+func resetDefaultsFromConfig(ctx context.Context, encryptionAtRestRSCurrent, encryptionAtRestRSNew, encryptionAtRestRSConfig *tfEncryptionAtRestRSModel) {
```

Comment:
I understand, I've added comments inside the handlers to make it more clear, I'm happy to add additional comments in this method as well.
 I faced some issues in this area (resetting values) with this resource because certain values are not returned by the API on Create/Update which need to be taken up from the config to keep it consistent with the state, otherwise the Framework throws an error for inconsistency.
Another scenario is that the API returns all blocks whether or not the user set it, in that scenario we need to unset the unused blocks based on what is in the config otherwise we run into the same inconsistency issue.

The idea is that during Read() (also called during Import where there's no user-defined config) we don't have the Config object available so we use the state/plan instead for this, but for Create/Update we have access to the Config object which we then use for this purpose.

Reply:
added additional comments to make it more clear

Reply:
Thanks for the comments in the code, this helps a lot. 
For reference, in alert configuration I faced a similar issue were I had certain attributes that where not returned by the API and had to be defined from existing state/config to avoid errors. The following function https://github.com/mongodb/terraform-provider-mongodbatlas/blob/CLOUDP-189585-plugin-framework-migration/mongodbatlas/fw_resource_mongodbatlas_alert_configuration.go#L641 was defined which receives only a single tfModel and receives the terraform state or plan depending on the context it is invoked:
> Create -> pass the Plan // planned state for the resource.
> Read -> pass the State // current state for the resource.
> Update -> pass the Plan // planned state for the resource.
> Import -> pass empty state as no existing state or plan is available.

You might have additional edge cases that justify receiving 2 models, but thought it would be worth sharing.

Reply:
@AgustinBettati yes, thank you for sharing this! You're right, for encryption at rest there are additional edge cases as listed above. The initial approach here was to use only a single other model but due to above cases, ended up having to use 2. Also, was trying to keep any resetting logic in a single method for all CRUD handlers.

Reply:
Okay, thanks for the follow up!